### ISSUE TITLE
Incorrect Line Number Reporting when Importing scienceplots

### ISSUE THREAD

**Issue by cospeng** [astral-sh/ruff#21675]
### Summary

I am encountering an issue with Ruff when importing the scienceplots library in my Python code. Specifically, when I import scienceplots, the line number in the error reports appears to be incorrect. The error is reported on the line above the actual line where the problem occurs.

**Steps to Reproduce:**

Create a Python script with the following content:

```
import scienceplots
a == 9
```

Run Ruff static analysis on the script.

The error will be reported on the line 1, even though the actual error occurs on the line 2.

**Expected Behavior:**

The error should be reported on the actual line where the issue occurs, without any discrepancy in the reported line number.

**Possible Cause:**

The likely reason for the error is that the imported scienceplots library isn't explicitly used in the code. I tried replacing it with other libraries, such as numpy—importing it without actually calling or using it—and encountered the same issue.

**Environment:**

Ruff version: 0.14.6

Python version: 3.13

Operating system:windows11 vim(ale+ruff)

### Version

0.14.6

**Comment by MichaReiser** [comment]
Can you tell me more how you run Ruff? The line numbers all look correct to me:

```
D100 Missing docstring in public module
--> test.py:1:1

I001 [*] Import block is un-sorted or un-formatted
 --> test.py:1:1
  |
1 | import scienceplots
  | ^^^^^^^^^^^^^^^^^^^
2 | a == 9
  |
help: Organize imports

F401 [*] `scienceplots` imported but unused
 --> test.py:1:8
  |
1 | import scienceplots
  |        ^^^^^^^^^^^^
2 | a == 9
  |
help: Remove unused import: `scienceplots`

B015 Pointless comparison. Did you mean to assign a value? Otherwise, prepend `assert` or remove it.
 --> test.py:2:1
  |
1 | import scienceplots
2 | a == 9
  | ^^^^^^
  |

F821 Undefined name `a`
 --> test.py:2:1
  |
1 | import scienceplots
2 | a == 9
  | ^
  |

PLR2004 Magic value used in comparison, consider replacing `9` with a constant variable
 --> test.py:2:6
  |
1 | import scienceplots
2 | a == 9
  |      ^
  |

Found 6 errors.
[*] 2 fixable with the `--fix` option.
```

**Comment by cospeng** [comment]
Thanks for the quick response!

I’m using **Ruff through ALE (Asynchronous Lint Engine) in Vim on Windows**, not via the command line. The line number issue **only appears when I enable `--fix` in ALE’s Ruff options**. When I comment out that line, all diagnostics show correct line numbers.

Here’s my relevant ALE configuration:

```vim
let g:ale_linters = {'python': ['ruff']}
let g:ale_fixers  = {'python': ['ruff']}
let g:ale_fix_on_save = 1
let g:ale_python_ruff_options = '--fix'   " ← this triggers the issue
```

From `:ALEInfo`, here’s the actual command ALE runs:

```text
ruff check -q --no-fix --fix --output-format json-lines --stdin-filename C:\Users\pyxuan\Documents\Python\todo\test.py - < C:\Users\pyxuan\AppData\Local\Temp\VMQ7276.tmp\test.py
```

And the output was:

```json
a == 1
{"cell":null,"code":"F821","end_location":{"column":2,"row":1},"filename":"C:\\Users\\..\\test.py","fix":null,"location":{"column":1,"row":1},"message":"Undefined name `a`","noqa_row":1,"url":"https://docs.astral.sh/ruff/rules/undefined-name"}
```

Notice that **the content sent to Ruff via stdin is only `a == 1`** — meaning the `import scienceplots` line was already removed (likely because Ruff fixed the unused import before linting). So Ruff correctly reports the error on `row: 1` **relative to the modified input**.

However, in my **original file**, that line is actually **line 2**, so ALE displays the diagnostic on the wrong line in the editor.

Since running `ruff check test.py` directly in PowerShell shows correct line numbers, this seems to be an **integration issue between ALE and Ruff in `--fix` mode on Windows**, possibly related to how ALE pipes the buffer content after applying fixes.

Thanks again for your help! Let me know if you’d still like more details (e.g., full file content, ALE logs).

And if this turns out to be relevant to ALE’s side, I’m happy to open an issue there too.

**Comment by MichaReiser** [comment]
> Notice that the content sent to Ruff via stdin is only a == 1 — meaning the import scienceplots line was already removed (likely because Ruff fixed the unused import before linting). So Ruff correctly reports the error on row: 1 relative to the modified input.

Interesting, why does ale only sent the second line to Ruff rather than the entire document? 

I don't see how Ruff should know about the correct line numbers if it doesn't receive the entire document. I think ALE either has to send the entire buffer or patch up Ruff's line numbers and convert them from relative to absolute line numbers again

**Comment by cospeng** [comment]
Thanks for the help and clarification.

After further investigation, I can confirm that **Ruff is reporting line numbers correctly** based on the input it receives. The issue arises because **ALE applies fixes first and then lints the *modified* buffer**, which no longer matches the original file structure.

In my case:
- Original file had unused imports (e.g., `import scienceplots`);
- ALE ran `ruff --fix`, which removed those lines;
- It then passed the **fixed (shorter) content** to Ruff for linting;
- Ruff correctly reported errors relative to that input (e.g., `row: 1`);
- But ALE displayed those diagnostics against the original buffer without remapping line numbers.

So the root cause lies in **how ALE integrates Ruff in `--fix` mode**, not in Ruff itself.  
I’ve worked around it by configuring ALE to **not pass `--fix` to the linter**, only to the fixer:

```vim
let g:ale_python_ruff_options = ''   " linter: no --fix
let g:ale_fixers = {'python': ['ruff']}
let g:ale_fix_on_save = 1
```

This ensures linting happens on the full original buffer, preserving correct line numbers.

Appreciate your support — feel free to close this issue!
